ui/app/scripts/controllers/newDeploymentCtrl.js

```diff
@@ -303,40 +344,168 @@ angular.module('apollo')
         };
 
        $scope.toggleSelectedEnvironment = function(environment) {
-            var index = $scope.selectedEnvironments.indexOf(environment);
+            var index = $scope.selectedEnvironmentsAndStacks.indexOf(environment);
             if (index > -1) {
-                $scope.selectedEnvironments.splice(index, 1);
+                $scope.selectedEnvironmentsAndStacks.splice(index, 1);
+                if(environment.isStack == true) {
+                    environment.environments.forEach(function(environmentId) {
+                         var environmentInStack = allEnvironments.find(x => x.id === environmentId);
+                         var index = $scope.selectedEnvironments.indexOf(environmentInStack);
+                         $scope.selectedEnvironments.splice(index, 1);
+                    });
+                }
+                var index = $scope.selectedEnvironments.indexOf(environment);
+                if(index > -1) {
+                    $scope.selectedEnvironments.splice(index, 1);
+                }
             }
             else {
-                if ($scope.isGroupDeployment) {
-                    $scope.selectedEnvironments = [];
+                if(environment.isStack == true) {
+                    var stack = environment;
+                    var environmentsNames = [];
+                    $scope.selectedEnvironmentsAndStacks.push(stack);
+                    stack.environments.forEach(function(environmentId) {
+                        var index = $scope.selectedEnvironments.indexOf(environment);
+                        if (index > -1) {
```

Comment:
If I select an environment, and then a stack that this environment is part of - won't this remove the selection for me?
There is no sense in this part IMO

Reply:
that way I'm not adding twice the same environment

Reply:
I still don't follow. This is the addition toggle flow, and you are protecting dups in line 375, no?